dask_image/ndinterp/__init__.py

```diff
@@ -0,0 +1,165 @@
+# -*- coding: utf-8 -*-
+
+
+import numpy as np
+import dask.array as da
+from scipy import ndimage
+import warnings
+
+
+__all__ = [
+    "affine_transform",
+]
+
+
+def affine_transform(
+        image,
+        matrix,
+        offset=0.0,
+        output_shape=None,
+        order=1,
+        output_chunks=None,
+        **kwargs
+):
+    """Apply an affine transform using Dask.
+    `ndimage.affine_transformation` for Dask Arrays. For every
+    output chunk, only the slice containing the relevant part
+    of the image is passed on to `ndimage.affine_transformation`.
+
+    Notes
+    -----
+        Differences to `ndimage.affine_transformation`:
+        - currently, prefiltering is not supported
+          (affecting the output in case of interpolation `order > 1`)
+        - default order is 1
+        - currently, matrix is expected to have shape (ndim,) or (ndim, ndim)
+          (because offset needs to be set for each chunk)
```

Comment:
But you can split up the matrix _internally_ to the function, as is done in `scipy.ndimage`, right? So you could still support (ndim+1, ndim+1), then, as soon as you enter the function, you rip it up into an (ndim, ndim) matrix and (ndim,) offsets.

Reply:
Yes, the parameters could be processed in the function. I guess at first I didn't want to copy this `ndimage` code that could be subject to changes:

    matrix = numpy.asarray(matrix, dtype=numpy.float64)
    if matrix.ndim not in [1, 2] or matrix.shape[0] < 1:
        raise RuntimeError('no proper affine matrix provided')
    if (matrix.ndim == 2 and matrix.shape[1] == input.ndim + 1 and
            (matrix.shape[0] in [input.ndim, input.ndim + 1])):
        if matrix.shape[0] == input.ndim + 1:
            exptd = [0] * input.ndim + [1]
            if not numpy.all(matrix[input.ndim] == exptd):
                msg = ('Expected homogeneous transformation matrix with '
                       'shape %s for image shape %s, but bottom row was '
                       'not equal to %s' % (matrix.shape, input.shape, exptd))
                raise ValueError(msg)
        # assume input is homogeneous coordinate transformation matrix
        offset = matrix[:input.ndim, input.ndim]
        matrix = matrix[:input.ndim, :input.ndim]
    if matrix.shape[0] != input.ndim:
        raise RuntimeError('affine matrix has wrong number of rows')
    if matrix.ndim == 2 and matrix.shape[1] != output.ndim:
        raise RuntimeError('affine matrix has wrong number of columns')
    if not matrix.flags.contiguous:
        matrix = matrix.copy()

But I agree with you that this is an important part of the function's interface and not being flexible with the parameter format could represent an obstacle to some use cases / users.

It would be great if the above code in `ndimage` was outsourced into a different function, to be reused here. But in the meantime it probably makes sense to just copy it here.

Reply:
> I guess at first I didn't want to copy this ndimage code that could be subject to changes:

I literally wrote that exact code so that hurt. 🤣🤣🤣

> in the meantime it probably makes sense to just copy it here.

I agree. 😝 

> It would be great if the above code in ndimage was outsourced into a different function, to be reused here.

I think SciPy would be reluctant to add such a function to their public API anyway. So it would be a private function which we probably would want to copy!

Reply:
> I literally wrote that exact code so that hurt. 🤣🤣🤣

Haha my statement referred to potentially duplicating it rather than the code itself 😁😂👍.

>I think SciPy would be reluctant to add such a function to their public API anyway. So it would be a private function which we probably would want to copy!

You're right, no need for a new function.

I think I found a good solution now (see recent commit). At first, an `ndimage.affine_transform` test run on a dummy array checks that parameters are valid (using your code!). In a second step, only the two relevant parameter formats that overwrite the offset are handled. This should give full support for all the formats and avoids testing the duplicated code :)